src/indexing.jl

```diff
@@ -70,44 +28,63 @@ end
     newaxes = Expr[]
     for d=1:lastnonscalar-droplastaxis
         if I[d] <: AxisArray
+            # Indexing with an AxisArray joins the axis names
             idxnames = axisnames(I[d])
             for i=1:ndims(I[d])
                 push!(newaxes, :($(Axis{Symbol(names[d], "_", idxnames[i])})(I[$d].axes[$i].val)))
             end
-        elseif I[d] <: Idx
-            push!(newaxes, :($(Axis{names[d]})(A.axes[$d].val[J[$d]])))
+        elseif I[d] <: Real
+        elseif I[d] <: Union{AbstractVector,Colon}
+            push!(newaxes, :($(Axis{names[d]})(A.axes[$d].val[Base.to_index(I[$d])])))
         elseif I[d] <: AbstractArray
             for i=1:ndims(I[d])
+                # When we index with non-vector arrays, we *add* dimensions.
                 push!(newaxes, :($(Axis{Symbol(names[d], "_", i)})(1:size(I[$d], $i))))
             end
         end
     end
     quote
-        # First copy the data using scalar indexing - an adaptation of Base
-        checkbounds(A, I...)
-        J = Base.to_indexes($(Isplat...))
-        sz = Base.index_shape(A, J...)
-        idx_lens = Base.index_lengths(A, J...)
-        src = A.data
-        dest = similar(A.data, sz)
-        D = eachindex(dest)
-        Ds = start(D)
-        Base.Cartesian.@nloops $N i d->(1:idx_lens[d]) d->(@inbounds j_d = J[d][i_d]) begin
-            d, Ds = next(D, Ds)
-            v = Base.Cartesian.@ncall $N unsafe_getindex src j
-            unsafe_setindex!(dest, v, d)
-        end
-        # And now create the AxisArray:
-        AxisArray(dest, $(newaxes...))
+        ($(newaxes...),)
     end
 end
 
+@inline function Base.getindex(A::AxisArray, idxs::Idx...)
+    AxisArray(A.data[idxs...], reaxis(A, idxs...))
+end
+
+# To resolve ambiguities, we need several definitions
+@inline function Base.view{T,N}(A::AxisArray{T,N}, idxs::Vararg{Idx,N})
+    AxisArray(view(A.data, idxs...), reaxis(A, idxs...))
+end
+function Base.view(A::AxisArray, idx::Idx)
+    AxisArray(view(A.data, idx), reaxis(A, idx))
+end
+@inline function Base.view{N}(A::AxisArray, idxs::Vararg{Idx,N})
+    # this should eventually be deleted, see julia #14770
+    AxisArray(view(A.data, idxs...), reaxis(A, idxs...))
+end
+
 # Setindex is so much simpler. Just assign it to the data:
-Base.setindex!(A::AxisArray, v, idxs::Idx...) = (A.data[idxs...] = v)
+@inline Base.setindex!(A::AxisArray, v, idxs::Idx...) = (A.data[idxs...] = v)
 
 ### Fancier indexing capabilities provided only by AxisArrays ###
-Base.getindex(A::AxisArray, idxs...) = A[to_index(A,idxs...)...]
-Base.setindex!(A::AxisArray, v, idxs...) = (A[to_index(A,idxs...)...] = v)
+@inline Base.getindex(A::AxisArray, idxs...) = A[to_index(A,idxs...)...]
+@inline Base.setindex!(A::AxisArray, v, idxs...) = (A[to_index(A,idxs...)...] = v)
+# Deal with lots of ambiguities here
```

Comment:
We should probably do the same sort of thing with `view` that we did for `getindex`: only define `view(::T, ::Any...)` in the standard library.  I wish it'd be easier to specify "fallback" methods.